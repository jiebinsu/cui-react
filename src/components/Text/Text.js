import React from "react";
import PropTypes from "prop-types";
import classNames from "../../utils/classNames";
import "./text.scss";

const variantList = {
  mega: "coop-t-h-mega",
  lead: "coop-t-lead-p",
  regular: "coop-t-regular",
  medium: "coop-t-medium",
  bold: "coop-t-bold",
};

const Text = ({ className, variant, type, href, children, ...props }) => {
  const TagName = type;
  const variantClassName = classNames(variantList[variant], [className]);

  const tagAttributes = {
    className: variantClassName,
    ...props,
  };

  if (type === "a") {
    const linkHref = href || tagAttributes.href;
    tagAttributes.href = linkHref;
  }

  return <TagName {...tagAttributes}>{children}</TagName>;
};

Text.defaultProps = {
  type: "p",
  variant: "regular",
  className: null,
  href: null,
  children: null,
};

Text.propTypes = {
  className: PropTypes.string,
  variant: PropTypes.oneOf(Object.keys(variantList)),
  type: PropTypes.oneOf(["h1", "h2", "h3", "h4", "h5", "h6", "p", "a"]),
  href: PropTypes.string,
  children: PropTypes.node,
};

export default Text;
