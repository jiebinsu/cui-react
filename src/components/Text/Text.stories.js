import React from "react";
import Text from "./Text";

export default {
  title: "Coop UI/Text",
  component: Text,
};

export const Basic = (args) => <Text {...args} />;
Basic.args = { children: "Label" };

export const All = () => (
  <>
    <Text variant="mega">Mega</Text>
    <Text variant="lead">Lead</Text>
    <Text variant="regular">Regular</Text>
    <Text size="medium">Medium</Text>
    <Text variant="bold">Bold</Text>
  </>
);
