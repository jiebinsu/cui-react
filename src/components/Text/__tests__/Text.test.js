import React from "react";
import { mount } from "enzyme";
import renderer from "react-test-renderer";
import Text from "../index";

describe("Text", () => {
  it("should render P element in the default", () => {
    const wrapper = mount(<Text>test-value</Text>);
    expect(wrapper.find("p").length).toBe(1);
    expect(() => wrapper.unmount()).not.toThrow();
  });

  it("should work with all variants", () => {
    const Variants = () => (
      <div>
        <Text variant="mega">Mega</Text>
        <Text variant="lead">Lead</Text>
        <Text variant="regular">Regular</Text>
        <Text size="medium">Medium</Text>
        <Text variant="bold">Bold</Text>
      </div>
    );

    const wrapper = mount(<Variants />);
    expect(() => wrapper.unmount()).not.toThrow();

    const tree = renderer.create(<Variants />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("the specified element should be rendered", () => {
    const elements = ["h1", "h2", "h3", "h4", "h5", "h6", "p", "a"];
    const Elements = () => (
      <div>
        {elements.map((el, index) => {
          const key = `${el}-${index}`;
          return <Text key={key}>test-value</Text>;
        })}
      </div>
    );
    const wrapper = mount(<Elements />);
    expect(() => wrapper.unmount()).not.toThrow();

    const tree = renderer.create(<Elements />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
