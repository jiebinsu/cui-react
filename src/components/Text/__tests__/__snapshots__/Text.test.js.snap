// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text should work with all variants 1`] = `
<div>
  <p
    className="coop-t-h-mega"
  >
    Mega
  </p>
  <p
    className="coop-t-lead-p"
  >
    Lead
  </p>
  <p
    className="coop-t-regular"
  >
    Regular
  </p>
  <p
    className="coop-t-regular"
    size="medium"
  >
    Medium
  </p>
  <p
    className="coop-t-bold"
  >
    Bold
  </p>
</div>
`;

exports[`Text the specified element should be rendered 1`] = `
<div>
  <p
    className="coop-t-regular"
  >
    test-value
  </p>
  <p
    className="coop-t-regular"
  >
    test-value
  </p>
  <p
    className="coop-t-regular"
  >
    test-value
  </p>
  <p
    className="coop-t-regular"
  >
    test-value
  </p>
  <p
    className="coop-t-regular"
  >
    test-value
  </p>
  <p
    className="coop-t-regular"
  >
    test-value
  </p>
  <p
    className="coop-t-regular"
  >
    test-value
  </p>
  <p
    className="coop-t-regular"
  >
    test-value
  </p>
</div>
`;
